### ISSUE TITLE
amazon-ebs: cannot suppress block device mapping with "no_device"

### ISSUE THREAD

**Issue by rvandegrift** [mitchellh/packer#2398]
With packer 0.8.1 and the amazon-ebs builder, I cannot suppress block device mappings.  The resulting AMI cannot be deployed.  This works correctly in 0.7.5.  Here is an example template:

```javascript
{
  "variables": {
    "aws_subnet_id": "",
    "aws_vpc_id": ""
  },
  "builders": [
    {
      "ami_block_device_mappings": [
        {
          "delete_on_termination": true,
          "volume_type": "gp2",
          "device_name": "/dev/sda1"
        },
        {
          "device_name": "/dev/sdb",
          "no_device": true
        },
        {
          "device_name": "/dev/sdc",
          "no_device": true
        }
      ],
      "type": "amazon-ebs",
      "region": "us-west-2",
      "vpc_id": "{{user `aws_vpc_id`}}",
      "subnet_id": "{{user `aws_subnet_id`}}",
      "source_ami": "ami-5189a661",
      "instance_type": "t2.micro",
      "ssh_username": "ubuntu",
      "ami_name": "test at {{timestamp}}"
    }
  ]
}
```

With 0.7.5, the resulting block device mapping looks like this:
```javascript
            "BlockDeviceMappings": [
                {
                    "Ebs": {
                        "SnapshotId": "snap-220de472",
                        "VolumeType": "gp2",
                        "DeleteOnTermination": true,
                        "Encrypted": false,
                        "VolumeSize": 8
                    },
                    "DeviceName": "/dev/sda1"
                }
            ],
```

With 0.8.1, the result looks like this:
```javascript
            "BlockDeviceMappings": [
                {
                    "Ebs": {
                        "DeleteOnTermination": true,
                        "Encrypted": false,
                        "VolumeType": "gp2",
                        "VolumeSize": 8,
                        "SnapshotId": "snap-3062bb63"
                    },
                    "DeviceName": "/dev/sda1"
                },
                {
                    "Ebs": {
                        "DeleteOnTermination": false,
                        "Encrypted": false,
                        "VolumeType": "standard",
                        "VolumeSize": 0
                    },
                    "DeviceName": "/dev/sdb"
                },
                {
                    "Ebs": {
                        "DeleteOnTermination": false,
                        "Encrypted": false,
                        "VolumeType": "standard",
                        "VolumeSize": 0
                    },
                    "DeviceName": "/dev/sdc"
                }
            ],
```

Neither /dev/sdb nor /dev/sdc is suppressed.  Since they have size 0 and no associated snapshot, this AMI cannot be deployed.

**Comment by catsby** [comment]
Hey @rvandegrift – thanks for opening this issue. I spent some time trying to dig into this yesterday but wasn't able to crack it... I *think* this is an upstream issue with the AWS SDK or the API. I've reproduced it outside of Packer and then opened an issue on the aws-sdk-go repo:

- https://github.com/aws/aws-sdk-go/issues/307

I'll try to keep this issue updated with what happens there. Thanks again

**Comment by cbednarski** [comment]
I'm pulling this from 0.8.2 since we're waiting on upstream.

**Comment by tb3088** [comment]
This is very much a Packer problem. It's mishandling booleans for starters (ie. it doesn't handle them at all - only deals with strings despite having the a proper function call to figure out datatypes) but it's not up to aws-sdk-go to accomodate "malformed" parameters handed to it by Packer.

If the Packer config is "no_device : 'true'" then it is an error on Packer's part to even submit the stanza to aws-sdk-go libraries and furthermore it's up to Packer to unwrap it's "strings" into their proper data types before calling the upstream library.

**Comment by mitchellh** [comment]
@tb3088 Without looking into the detail more, I do want to note that in Packer we do decode it into the proper data type from the config. We allow the weak typing in the config itself (strings), but the _moment_ it is decoded it goes directly to a real data type using mapstructure, which we also use for Terraform, Consul, Vault, and more. It is highly stable project :)

**Comment by mitchellh** [comment]
@tb3088 Also, we've reproduced this outside of Packer with just aws-sdk-go on its own (with proper data types), and we've also been in discussion and confirmation with upstream of the problem. This isn't a typing problem as far as we can tell.

**Comment by tb3088** [comment]
> we do decode it into the proper data type from the config.
Sorry, yes. I meant to clarify about not handling data types in the config itself. It really irks when strings and only strings are tolerated.

**Comment by WyseNynja** [comment]
Any workarounds while we wait for the fix?

**Comment by lightpriest** [comment]
I've stumbled upon this problem as well. I found that if you only include DeviceName and NoDevice, they are properly being removed from the launch specification *and* from the created AMI. So if this:

```json
"ami_block_device_mappings|launch_block_device_mappings": [
  {"device_name": "/dev/sdb", "no_device": true},
  {"device_name": "/dev/sdc", "no_device": true}
]
```

Is turned into this:

```go
BlockDevices: [{
  DeviceName: "/dev/sdb",
  NoDevice: ""
} {
  DeviceName: "/dev/sdc",
  NoDevice: ""
}]
```

It creates the expected mapping. Currently it is mapped like in the first post and even if you use set virtual_name to ephemeralN, the existence of "VirtualName" in the final mapping ruins it. Unfortunately, I couldn't work-around this without patching `builder/common/amazon/block_device.go`.

I didn't check if this is a real upstream (aws-sdk) issue or just a quirk in its API. I also haven't managed to run a clean `make test` without it failing on `windows-restart` provisioner so I wasn't sure if I should submit a PR. Is there anyway to skip a test?

Just in case, this is diff:

```diff
diff --git a/builder/amazon/common/block_device.go b/builder/amazon/common/block_device.go
index 0947388..42ef9e2 100644
--- a/builder/amazon/common/block_device.go
+++ b/builder/amazon/common/block_device.go
@@ -49,16 +49,18 @@ func buildBlockDevices(b []BlockDevice) []*ec2.BlockDeviceMapping {
                }
 
                mapping := &ec2.BlockDeviceMapping{
-                       DeviceName:  aws.String(blockDevice.DeviceName),
-                       VirtualName: aws.String(blockDevice.VirtualName),
+                       DeviceName: aws.String(blockDevice.DeviceName),
                }
 
-               if !strings.HasPrefix(blockDevice.VirtualName, "ephemeral") {
-                       mapping.Ebs = ebsBlockDevice
+               if blockDevice.VirtualName != "" {
+                       mapping.VirtualName = aws.String(blockDevice.VirtualName)
                }
 
+               // No need to specify Ebs if this is a NoDevice
                if blockDevice.NoDevice {
                        mapping.NoDevice = aws.String("")
+               } else if !strings.HasPrefix(blockDevice.VirtualName, "ephemeral") {
+                       mapping.Ebs = ebsBlockDevice
                }
 
                blockDevices = append(blockDevices, mapping)
```

**Comment by tb3088** [comment]
Seems simpler to wrap 'blockDevices = append(blockDevices, mapping)' in an IF that doesn't fire when blockDevice.NoDevice is TRUE

**Comment by lightpriest** [comment]
@tb3088 you're right, this diff isn't PR quality, this section probably needs some small refactoring. I think removing the mapping from blockDevices would not have the desired effect for all cases (depending on the source AMI, iirc).

I guess it could  be as follows:

```
if no_device
  create mapping with NoDevice and DeviceName only
else if VirtualName begins with ephemeralN
  create mapping with VirtualName and DeviceName only
otherwise
  create explicit* mapping with Ebs

* A mapping that doesn't include keys by default, only what the user has specified.
```

**Comment by knope** [comment]
#### perhaps there is a better place to scream:
#  'HALP!!!!'
   *(am noob and perhaps I'm completely misunderstanding the uses of the 'device_name'; how ever I do wonder if i'm hitting some form of this bug; as understanding from docs is that this is the correct place to assign a name such that '/dev/cowX' )*

###### but this looks like the most recently updated version of this problem that I can't seem to get around...

## (packer version: 0.8.5 )

### given something like the following (inside 'builders: [ ]' ):

```
      "ami_block_device_mappings": [ {
            "device_name": "sda1",
            "volume_size": "13",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        },
        {
            "device_name": "sdb1",
            "volume_size": "9",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        }
        ],
        "launch_block_device_mappings": [ {
            "device_name": "sda1",
            "volume_size": "13",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        },
        {
            "device_name": "sdb1",
            "volume_size": "9",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        } ]}

```
### or even:

```
      "ami_block_device_mappings": [ {
            "device_name": "/dev/sda1",
            "volume_size": "13",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        },
        {
            "device_name": "/dev/sdb1",
            "volume_size": "9",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        }
        ],
        "launch_block_device_mappings": [ {
            "device_name": "/dev/sda1",
            "volume_size": "13",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        },
        {
            "device_name": "/dev/sdb1",
            "volume_size": "9",
            "volume_type": "ssd",
            "delete_on_termination": "true"
        } ]}
```
________________________

### both return:

```
==> amazon-ebs: Error launching source instance: InvalidBlockDeviceMapping: Invalid device name sdb1
==> amazon-ebs: 	status code: 400, request id: []
```

**Comment by jsternberg** [comment]
We're running into this issue too. Packer 0.7.5 previously worked for us and we can't upgrade to a new version because of this. Contrary to what was mentioned right above me, this config does not work correctly:

```
      "ami_block_device_mappings": [
        {
          "device_name": "/dev/sdb",
          "no_device": true
        }
      ],
```

From a look at the code, this appears to be an issue in Packer.

```
        if !strings.HasPrefix(blockDevice.VirtualName, "ephemeral") {
            mapping.Ebs = ebsBlockDevice
        }

        if blockDevice.NoDevice {
            mapping.NoDevice = aws.String("")
        }
```

It looks like the first line there should be changed to:

```
if !(blockDevice.NoDevice || strings.HasPrefix(blockDevice.VirtualName, "ephemeral")) {
    mapping.Ebs = ebsBlockDevice
}
```

Otherwise it's just going to pass the wrong parameters in. I'll give it a test and send a PR if it works.

**Comment by jsternberg** [comment]
I just submitted a PR with my change. I didn't pay attention to @lightpriest's change before, but it looks like his does the same thing as mine and does the same thing. The only thing I would change is if `VirtualName` is specified at all to use it for an ephemeral volume. It could be misleading if somebody specifies `VirtualName` with a wrong name and it tries to create an EBS device instead.

**Comment by markpeek** [comment]
Fixed via merge of #2672.
